File: AIペアプログラミング実践編_ソースコード/practical-csharp/session-4/Ans/Inventory.Api/Repositories/ProductRepository.cs
namespace Inventory.Api.Repositories;

using Inventory.Api.Data;
using Inventory.Api.Entities;

public class ProductRepository : IProductRepository
{
    private readonly InventoryDbContext _context;

    public ProductRepository(InventoryDbContext context)
    {
        _context = context;
    }

    public Product? FindById(int id)
    {
        return _context.Products.Find(id);
    }

    public List<Product> FindAll()
    {
        return _context.Products.ToList();
    }

    public Product Save(Product product)
    {
        if (product.Id == 0)
        {
            _context.Products.Add(product);
        }
        else
        {
            _context.Products.Update(product);
        }
        _context.SaveChanges();
        return product;
    }

    public void Delete(int id)
    {
        var product = _context.Products.Find(id);
        if (product != null)
        {
            _context.Products.Remove(product);
            _context.SaveChanges();
        }
    }

    public List<Product> Search(string? name)
    {
        var query = _context.Products.AsQueryable();

        if (!string.IsNullOrEmpty(name))
        {
            query = query.Where(p => p.Name.Contains(name));
        }

        return query.ToList();
    }
}

File: AIペアプログラミング実践編_ソースコード/practical-python/session-4/Ans/inventory/repositories/product_repository.py
"""商品リポジトリ"""
from inventory.models.product import Product


class ProductRepository:
    """商品リポジトリ"""

    def save(self, product: Product) -> Product:
        """商品を保存"""
        product.save()
        return product

    def find_by_id(self, product_id: int) -> Product | None:
        """ID で商品を取得"""
        return Product.objects.filter(id=product_id).first()

    def find_all(self):
        """全商品を取得"""
        return list(Product.objects.all())

    def find_by_name_containing(self, name: str):
        """商品名で部分一致検索"""
        return list(Product.objects.filter(name__icontains=name))
